sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -111,6 +113,24 @@ object DateTimeUtils {
     computedTimeZones.computeIfAbsent(timeZoneId, computeTimeZone)
   }
 
+  private val threadLocalComputedCalendarsMap =
+    new ThreadLocal[mutable.Map[TimeZone, (Calendar, Long)]] {
+      override def initialValue(): mutable.Map[TimeZone, (Calendar, Long)] = {
+        mutable.Map[TimeZone, (Calendar, Long)]()
+      }
+    }
+
+  def getCalendar(timeZone: TimeZone): Calendar = {
+    val (c, timeInMillis) = threadLocalComputedCalendarsMap.get()
+      .getOrElseUpdate(timeZone, {
+        val c = Calendar.getInstance(timeZone)
+        (c, c.getTimeInMillis)
```

Comment:
When you get the calendar instance next time, isn't its time out of date?

Reply:
Please refer line 130 for this. Before returning the calendar instance, it is reinitialized to the time it was originally created.

Reply:
@viirya ^ Does that answer you question, or you mean something else?

Reply:
Isn't `timeInMillis` also stored when you first update this map entity? So next time you access this calendar, you just set it with the old `timeInMillis`. Isn't it?

Reply:
Yes, correct.